### PR TITLE
feat: atclient_get_sharedkey 

### LINKED ISSUE #135: at_c: atclient_get_sharedkey
Basic atProtocol get CRUD operation provided a Shared AtKey

- Use lookup: if it is shared with you, or llookup: if it is shared by you.
- Need to decrypt value with the correct AES Symmetric Encryption key

### PR DESCRIPTION
**Pull request by realvarx**
**- What I did**
- `atclient_get_sharedkey`
- `atclient_get_shared_by_me_with_other`
- `atclient_get_shared_by_other_with_me`

TODOS:

- [x] Manage dynamic memory free calls

**- How to verify it**

Manual test below

**- Description for the changelog**

Closes #135

### CODE REVIEW

**Comment by realvarx** [comment]
Reserved for manual test [ Dart ]

```
import 'dart:convert';
import 'dart:io';

import 'package:args/args.dart';
import 'package:at_client/at_client.dart';
import 'package:at_utils/at_utils.dart';
import 'package:crypto/crypto.dart';
import 'package:dart_playground/util.dart';

Future<void> main(List<String> arguments) async {
  // get atsign argument from ArgsParser
  final ArgParser argsParser = ArgParser();
  argsParser.addOption('sharedby',
      abbr: 'b', help: 'Owner of key', mandatory: true);
  argsParser.addOption('sharedwith',
      abbr: 'w', help: 'Recipient', mandatory: true);
  argsParser.addOption('name',
      abbr: 'k',
      help: 'Name of the AtKey',
      mandatory: false,
      defaultsTo: 'test');
  argsParser.addOption('value',
      abbr: 'v',
      help: 'Value',
      mandatory: false,
      defaultsTo: 'some encrypted value !!! 123');
  argsParser.addOption('namespace',
      abbr: 'n',
      help: 'Namespace',
      mandatory: false,
      defaultsTo: 'dart_playground');

  final ArgResults argResults;
  try {
    argResults = argsParser.parse(arguments);
  } catch (e) {
    print(argsParser.usage);
    return;
  }

  AtSignLogger.root_level = "shout";

  final String sharedByAtSign = argResults['sharedby'];
  final String sharedWithAtSign = argResults['sharedwith'];
  final String atKeyName = argResults['name'];
  final String atKeyValue = argResults['value'];
  final String namespace = argResults['namespace'];

  final AtClient atClient = await onboard(sharedByAtSign);

  AtKey atKey = AtKey()
        ..key = atKeyName
        ..sharedBy = sharedByAtSign
        ..sharedWith = sharedWithAtSign
        ..namespace = namespace // default namespace
      ;

  Metadata metadata = Metadata()
        ..ttl = 86400
        ..ccd = true
        ..ttr = 1000 * 1 * 60 // 1 minute
      ;

  atKey.metadata = metadata;

  var bytesKey = utf8.encode(atKey.toString());
  var digestKey = sha256.convert(bytesKey);
  
  var bytesValue = utf8.encode(atKeyValue);
  var digestValue = sha256.convert(bytesValue);

  print('Putting ${atKey.toString()} with value $atKeyValue');
  print('\tsha256digest ("key" : "value") => "$digestKey" : "$digestValue"');

  bool success = await atClient.put(atKey, atKeyValue,
      putRequestOptions: PutRequestOptions()..useRemoteAtServer = true);
  if (success) {
    print('Success');
    exit(0);
  } else {
    print('Failed');
    exit(1);
  }
}
```

Output:
```
C:\Users\Alvaro\Desktop\atsign\dart_playground>dart bin/make_sharedkey.dart -b "@expensiveferret" -w "@secondaryjackal" -k "test_sharedkey_001"                                                        
Putting @secondaryjackal:test_sharedkey_001.dart_playground@expensiveferret with value some encrypted value !!! 123
        sha256digest ("key" : "value") => "a90c5a61fad5e5fd64bd540940cd75d7346ff1449e5d0285359c33d26000753a" : "d38c816073fad5249b0143fe4588013a64a4f9d50090d9a8d4c97ad0534b2592"
Success
C:\Users\Alvaro\Desktop\atsign\dart_playground>    
```

(Forked from @JeremyTubongbanua dart_playground)

**Comment by realvarx** [comment]
Reserved for manual test [ C ]

https://github.com/atsign-foundation/at_c/blob/2724acf905277fe45745cc96633346cc6c9de937/packages/atclient/tests/noci_test_get_sharedkey.c#L1-L96

Output (sensitive data has been changed by 'xxxxx'):
```
test 1
    Start 1: noci_test_get_sharedkey

1: Test command: /home/realvarx/Desktop/at_c/packages/atclient/build/tests/noci_test_get_sharedkey
1: Working Directory: /home/realvarx/Desktop/at_c/packages/atclient/build/tests
1: Test timeout computed to be: 10000000
1:  [INFO]      connection |    SENT: "expensiveferret"
1:  [INFO]      connection |    RECV: "291ae081-0947-5525-b1a4-9dbc220b0db2.swarm0002.atsign.zone:6581"
1:  [INFO]      atclient | atclient_connection_connect: 0. Successfully connected to secondary
1:  [INFO]      connection |    SENT: "from:expensiveferret"
1:  [INFO]      connection |    RECV: "data:_35b7ef55-4b1e-4900-8243-c96b5b9805fd@expensiveferret:66d61fae-5e69-4675-86f2-461399832a4d"
1:  [INFO]      connection |    SENT: "pkam:xxxxx"
1:  [INFO]      connection |    RECV: "data:success"
1:  [DEBG]      noci_test_get_sharedkey | Authenticated
1:  [DEBG]      noci_test_get_sharedkey | Created shared key
1:  [DEBG]      noci_test_get_sharedkey | atkeystr.str (67): "@secondaryjackal:test_sharedkey_001.dart_playground@expensiveferret"
1:  [INFO]      connection |    SENT: "llookup:shared_key.secondaryjackal@expensiveferret"
1:  [INFO]      connection |    RECV: "data:xxxxx"
1:  [INFO]      connection |    SENT: "llookup:all:@secondaryjackal:test_sharedkey_001.dart_playground@expensiveferret"
1:  [INFO]      connection |    RECV: "data:{"key":"@secondaryjackal:test_sharedkey_001.dart_playground@expensiveferret","data":"xxxxx","metaData":{"createdBy":"@expensiveferret","updatedBy":"@expensiveferret","createdAt":"2024-03-19 17:51:20.980Z","updatedAt":"2024-03-19 17:51:20.980Z","expiresAt":"2024-03-19 17:52:47.380Z","refreshAt":"2024-03-20 10:31:20.980Z","status":"active","version":0,"ttl":86400,"ttr":60000,"ccd":true,"isBinary":false,"isEncrypted":false,"sharedKeyEnc":"xxxxx","pubKeyCS":"xxxxx","ivNonce":"xxxxx"}}"
1:  [DEBG]      noci_test_get_sharedkey | value.str (28): "some encrypted value !!! 123"
1/1 Test #1: noci_test_get_sharedkey ..........   Passed    2.61 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   2.61 sec
```

File: packages/atclient/tests/noci_test_get_sharedkey.c [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+#include <atclient/constants.h>
+#include <atclient/atclient.h>
+#include <atclient/atkey.h>
+#include <atclient/atsign.h>
+
+#define TAG "noci_test_get_sharedkey"
+
+#define ATSIGN "@expensiveferret"
+
+#define ATKEYS_FILE_PATH "/home/realvarx/.atsign/keys/@expensiveferret_key.atKeys"
```
Comment (JeremyTubongbanua) [inline packages/atclient/tests/noci_test_get_sharedkey.c]: Should we be merging tests like this that depend on keys locally?

Comment (realvarx) [reply packages/atclient/tests/noci_test_get_sharedkey.c]: We also have examples which depend on local keys. We could merge those tests (using the noci prefix for now, so we can adjust the variables and run them locally) and convert them to ci when we setup the secrets we need for them

Comment (JeremyTubongbanua) [reply packages/atclient/tests/noci_test_get_sharedkey.c]: I feel if that's the case, then it should live in `examples/`. It does not feel right to me to have tests that can't be ran in CI be in a `tests/` folder.

What do you think?

Comment (JeremyTubongbanua) [reply packages/atclient/tests/noci_test_get_sharedkey.c]: It also looks like to me you were using this for manual tests while you were developing. I did something similar with my `debug.sh` script [here](https://github.com/atsign-foundation/at_c/blob/trunk/examples/desktop/crud/debug.sh)

**Comment by JeremyTubongbanua** [comment]
To add on, we will definitely use this "noci" test code that you wrote. I think it's best to put it in examples/ (since it cannot live alone as a "test")

Then once we are ready, we can refer to the code here and write a robust functional/end-to-end test

**Comment by realvarx** [comment]
> It also looks like to me you were using this for manual tests while you were developing. I did something similar with my `debug.sh` script [here](https://github.com/atsign-foundation/at_c/blob/trunk/examples/desktop/crud/debug.sh)

Ohh I see `debug.sh` works similarly @JeremyTubongbanua . I hadnt seen that. Will leave here current commit ID for future references in case we need it, and move the test to `/examples/desktop/crud/`:
https://github.com/atsign-foundation/at_c/pull/149/commits/4aeeb80fc0d18ab1ec3736ba851729085217c698

**Review by JeremyTubongbanua — APPROVED** [review]

**Comment by JeremyTubongbanua** [comment]
Thanks for this work @realvarx

**Pull request merged**